### ISSUE TITLE
The panic branch of `Some(x).unwrap()` is not removed for some types

### ISSUE THREAD

**Issue by saethlin** [rust-lang/rust#93011]
Demos: https://godbolt.org/z/TMh3Krfc5

The motivating example is:
```rust
pub fn bad_vec(s: Vec<u8>) -> Vec<u8> {
    Some(s).expect("bad_vec")
}
```
```asm
example::bad_myvec:
        push    rax
        mov     rax, qword ptr [rsi]
        test    rax, rax
        je      .LBB5_1
        mov     qword ptr [rdi], rax
        movups  xmm0, xmmword ptr [rsi + 8]
        movups  xmmword ptr [rdi + 8], xmm0
        mov     rax, rdi
        pop     rcx
        ret
.LBB5_1:
        lea     rdi, [rip + .L__unnamed_5]
        lea     rdx, [rip + .L__unnamed_6]
        mov     esi, 9
        call    qword ptr [rip + core::option::expect_failed@GOTPCREL]
        ud2
```
It looks to me like the problem with `Vec` is that specifically the `nonnull` attribute does not survive being wrapped in two structs that each contain a second member. If either of `RawVec` or `Vec` is missing a `usize` member, the optimization works.

There is a missed `NonZero*` optimization in here that I found looking for types that fail to optimize. It's unclear to me if this was reported previously in https://github.com/rust-lang/rust/issues/49572, or if whatever fixes the problem with `Vec` will also fix the `NonZero*` example.

---
I think this is similar but not the same problem as seen in https://github.com/rust-lang/rust/issues/71257

**Comment by scottmcm** [comment]
You might be interested in #85646 (which came from #85133), which tried to make some progress here and found some LLVM issues with `match`ing on 2-variant enums (like the `Option` here).

nox was working on some codegen improvements that might help at some point: https://rust-lang.zulipchat.com/#narrow/stream/189540-t-compiler.2Fwg-mir-opt/topic/Discriminant.20computation.20yields.20bad.20output/near/264996641

**Comment by saethlin** [comment]
🤔 this is fixed by `DataflowConstProp`

**Comment by cjgillot** [comment]
The motivating example is handled by https://github.com/rust-lang/rust/pull/109597
Test: https://github.com/rust-lang/rust/blob/5a345b325b59370171d9a00c8f575fb177ead767/tests/mir-opt/gvn.rs#L148-L157
